File: tests/patent/test-monte-carlo.js
/**
 * © 2026 Heady™Systems Inc. PROPRIETARY AND CONFIDENTIAL.
 *
 * Tests: Monte Carlo Engine
 * Covers: runSimulation, distributions, confidence intervals, pipeline hooks,
 *         quickReadiness, scenario analysis, risk scoring.
 */

'use strict';

const assert = require('assert');
const {
  MonteCarloEngine,
  mulberry32,
  sampleUniform,
  sampleNormal,
  sampleTriangular,
  sample,
  wilsonInterval,
  descriptiveStats,
  scoreToGrade,
  PHI,
  RISK_GRADE,
  DISTRIBUTION,
  OUTCOME_THRESHOLDS,
} = require('../src/intelligence/monte-carlo-engine');

let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  ✓  ${name}`);
    passed++;
  } catch (err) {
    console.error(`  ✗  ${name}`);
    console.error(`       ${err.message}`);
    failed++;
  }
}

// ─── Constants ────────────────────────────────────────────────────────────────

console.log('\n── Constants ─────────────────────────────────────────────────────');

test('PHI equals 1.6180339887', () => {
  assert.strictEqual(PHI, 1.6180339887);
});

test('RISK_GRADE has GREEN, YELLOW, ORANGE, RED', () => {
  assert(RISK_GRADE.GREEN  === 'GREEN');
  assert(RISK_GRADE.YELLOW === 'YELLOW');
  assert(RISK_GRADE.ORANGE === 'ORANGE');
  assert(RISK_GRADE.RED    === 'RED');
});

test('OUTCOME_THRESHOLDS.SUCCESS_MAX is 0.30', () => {
  assert.strictEqual(OUTCOME_THRESHOLDS.SUCCESS_MAX, 0.30);
});

// ─── Mulberry32 PRNG ──────────────────────────────────────────────────────────

console.log('\n── Mulberry32 PRNG ───────────────────────────────────────────────');

test('mulberry32 returns values in [0, 1)', () => {
  const rand = mulberry32(42);
  for (let i = 0; i < 1000; i++) {
    const v = rand();
    assert(v >= 0 && v < 1, `Out of range: ${v}`);
  }
});

test('mulberry32 is deterministic with same seed', () => {
  const rand1 = mulberry32(123);
  const rand2 = mulberry32(123);
  for (let i = 0; i < 20; i++) {
    assert.strictEqual(rand1(), rand2());
  }
});

test('mulberry32 produces different sequences for different seeds', () => {
  const rand1 = mulberry32(1);
  const rand2 = mulberry32(2);
  let same = true;
  for (let i = 0; i < 10; i++) {
    if (rand1() !== rand2()) { same = false; break; }
  }
  assert(!same, 'Expected different sequences for different seeds');
});

// ─── Distribution Samplers ────────────────────────────────────────────────────

console.log('\n── Distribution Samplers ─────────────────────────────────────────');

test('sampleUniform returns values in [min, max]', () => {
  const rand = mulberry32(7);
  for (let i = 0; i < 1000; i++) {
    const v = sampleUniform(rand, 5, 10);
    assert(v >= 5 && v <= 10, `Out of range: ${v}`);
  }
});

test('sampleNormal produces values centred around mean', () => {
  const rand   = mulberry32(9);
  const values = [];
  for (let i = 0; i < 10_000; i++) values.push(sampleNormal(rand, 50, 10));
  const avg = values.reduce((a, b) => a + b, 0) / values.length;
  assert(Math.abs(avg - 50) < 1, `Expected mean ≈ 50, got ${avg.toFixed(2)}`);
});

test('sampleTriangular returns values in [min, max]', () => {
  const rand = mulberry32(11);
  for (let i = 0; i < 1000; i++) {
    const v = sampleTriangular(rand, 0, 0.5, 1);
    assert(v >= 0 && v <= 1, `Out of range: ${v}`);
  }
});

test('sample dispatches to correct distribution', () => {
  const rand = mulberry32(13);
  const u = sample(rand, DISTRIBUTION.UNIFORM,    { min: 0, max: 1 });
  const n = sample(rand, DISTRIBUTION.NORMAL,     { mean: 0, stddev: 1 });
  const t = sample(rand, DISTRIBUTION.TRIANGULAR, { min: 0, mode: 0.5, max: 1 });
  assert(typeof u === 'number');
  assert(typeof n === 'number');
  assert(typeof t === 'number');
});

// ─── Statistical Helpers ──────────────────────────────────────────────────────

console.log('\n── Statistical Helpers ───────────────────────────────────────────');

test('scoreToGrade maps ranges correctly', () => {
  assert.strictEqual(scoreToGrade(100), RISK_GRADE.GREEN);
  assert.strictEqual(scoreToGrade(80),  RISK_GRADE.GREEN);
  assert.strictEqual(scoreToGrade(79),  RISK_GRADE.YELLOW);
  assert.strictEqual(scoreToGrade(60),  RISK_GRADE.YELLOW);
  assert.strictEqual(scoreToGrade(59),  RISK_GRADE.ORANGE);
  assert.strictEqual(scoreToGrade(40),  RISK_GRADE.ORANGE);
  assert.strictEqual(scoreToGrade(39),  RISK_GRADE.RED);
  assert.strictEqual(scoreToGrade(0),   RISK_GRADE.RED);
});

test('wilsonInterval returns bounds in [0, 1]', () => {
  const { lower, upper } = wilsonInterval(0.1, 1000);
  assert(lower >= 0 && lower <= 1);
  assert(upper >= 0 && upper <= 1);
  assert(upper > lower);
});

test('wilsonInterval handles zero sample size', () => {
  const r = wilsonInterval(0, 0);
  assert.strictEqual(r.lower, 0);
  assert.strictEqual(r.upper, 1);
});

test('descriptiveStats computes correct mean and min/max', () => {
  const stats = descriptiveStats([1, 2, 3, 4, 5]);
  assert.strictEqual(stats.mean, 3.0);
  assert.strictEqual(stats.min,  1.0);
  assert.strictEqual(stats.max,  5.0);
});

test('descriptiveStats handles empty array', () => {
  const stats = descriptiveStats([]);
  assert.strictEqual(stats.mean, 0);
});

// ─── runSimulation ────────────────────────────────────────────────────────────

console.log('\n── runSimulation ─────────────────────────────────────────────────');

test('runSimulation returns required fields', () => {
  const eng = new MonteCarloEngine();
  const res = eng.runSimulation({ name: 'test', seed: 42 }, 100);
  assert(res.scenario === 'test');
  assert(typeof res.confidence   === 'number');
  assert(typeof res.failureRate  === 'number');
  assert(typeof res.successRate  === 'number');
  assert(typeof res.partialRate  === 'number');
  assert(res.outcomes);
  assert(res.confidenceBounds);
  assert(res.impactDistribution);
});

test('runSimulation outcomes sum to iterations', () => {
  const eng = new MonteCarloEngine();
  const res = eng.runSimulation({ seed: 1 }, 1000);
  const sum = res.outcomes.success + res.outcomes.partial + res.outcomes.failure;
  assert.strictEqual(sum, 1000);
});

test('runSimulation is deterministic with same seed', () => {
  const eng = new MonteCarloEngine();
  const r1  = eng.runSimulation({ name: 'det', seed: 777 }, 500);
  const r2  = eng.runSimulation({ name: 'det', seed: 777 }, 500);
  assert.strictEqual(r1.confidence,   r2.confidence);
  assert.strictEqual(r1.failureRate,  r2.failureRate);
});

test('runSimulation with high-probability risk factors reduces confidence', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.runSimulation({
    seed: 42,
    riskFactors: [
      { name: 'outage',   probability: 0.9, impact: 0.8 },
      { name: 'breach',   probability: 0.8, impact: 0.9 },
    ],
  }, 1000);
  assert(r.confidence < 50, `Expected low confidence, got ${r.confidence}`);
});

test('runSimulation with low-probability risks yields high confidence', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.runSimulation({
    seed: 99,
    riskFactors: [
      { name: 'minor', probability: 0.01, impact: 0.05 },
    ],
  }, 1000);
  assert(r.confidence > 80, `Expected high confidence, got ${r.confidence}`);
});

test('mitigation reduces effective impact', () => {
  const eng = new MonteCarloEngine();
  const without = eng.runSimulation({
    seed: 42,
    riskFactors: [{ name: 'risk', probability: 0.8, impact: 0.9 }],
  }, 1000);
  const withMit = eng.runSimulation({
    seed: 42,
    riskFactors: [{ name: 'risk', probability: 0.8, impact: 0.9, mitigation: 'firewall', mitigationReduction: 0.8 }],
  }, 1000);
  assert(withMit.confidence >= without.confidence, 'Mitigation should improve confidence');
});

test('runSimulation riskGrade is consistent with confidence', () => {
  const eng = new MonteCarloEngine();
  const res = eng.runSimulation({ seed: 42 }, 1000);
  assert.strictEqual(res.riskGrade, scoreToGrade(res.confidence));
});

test('runSimulation includes topMitigations list', () => {
  const eng = new MonteCarloEngine();
  const res = eng.runSimulation({
    seed: 42,
    riskFactors: [
      { name: 'r1', probability: 0.5, impact: 0.5, mitigation: 'patch' },
    ],
  }, 500);
  assert(Array.isArray(res.topMitigations));
});

test('runSimulation includes normal distribution support', () => {
  const eng = new MonteCarloEngine();
  const res = eng.runSimulation({
    seed: 42,
    riskFactors: [{
      name: 'normal-risk',
      probability: 0.5,
      impact: 0.3,
      distribution: DISTRIBUTION.NORMAL,
      distributionParams: { mean: 0.3, stddev: 0.1 },
    }],
  }, 500);
  assert(typeof res.confidence === 'number');
});

// ─── quickReadiness ───────────────────────────────────────────────────────────

console.log('\n── quickReadiness ────────────────────────────────────────────────');

test('quickReadiness returns GREEN for healthy signals', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.quickReadiness({
    errorRate: 0, lastDeploySuccess: true, cpuPressure: 0,
    memoryPressure: 0, serviceHealthRatio: 1, openIncidents: 0,
  });
  assert.strictEqual(r.grade, RISK_GRADE.GREEN);
  assert(r.score >= 80);
});

test('quickReadiness returns RED for degraded signals', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.quickReadiness({
    errorRate: 0.4, lastDeploySuccess: false, cpuPressure: 0.9,
    memoryPressure: 0.9, serviceHealthRatio: 0.2, openIncidents: 5,
  });
  assert(r.score < 60, `Expected low score, got ${r.score}`);
});

test('quickReadiness breakdown sums approximate to score', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.quickReadiness({});
  assert(r.breakdown);
  assert(typeof r.breakdown.errorScore  === 'number');
  assert(typeof r.breakdown.deployScore === 'number');
});

// ─── Scenario Analysis ────────────────────────────────────────────────────────

console.log('\n── Scenario Analysis ─────────────────────────────────────────────');

test('analyseScenarios returns comparison with best and worst', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.analyseScenarios([
    { name: 'optimistic', params: { seed: 1, riskFactors: [] }, iterations: 200 },
    { name: 'pessimistic', params: { seed: 2, riskFactors: [{ probability: 0.9, impact: 0.9 }] }, iterations: 200 },
  ]);
  assert(r.comparison.best);
  assert(r.comparison.worst);
  assert(typeof r.comparison.average === 'number');
});

// ─── Pipeline Hooks ───────────────────────────────────────────────────────────

console.log('\n── Pipeline Stage Integration ────────────────────────────────────');

test('registerPipelineHook is called after simulation for matching stage', () => {
  const eng   = new MonteCarloEngine();
  let hookCalled = false;
  eng.registerPipelineHook('risk-stage', (result) => { hookCalled = true; });
  eng.runSimulation({ name: 'with-hook', seed: 1, pipelineStage: 'risk-stage' }, 100);
  assert(hookCalled, 'Expected pipeline hook to be called');
});

test('clearPipelineHooks removes hooks for stage', () => {
  const eng = new MonteCarloEngine();
  let count = 0;
  eng.registerPipelineHook('stage-x', () => { count++; });
  eng.clearPipelineHooks('stage-x');
  eng.runSimulation({ pipelineStage: 'stage-x', seed: 1 }, 100);
  assert.strictEqual(count, 0, 'Hook should not be called after clearing');
});

// ─── Risk Scoring Utility ─────────────────────────────────────────────────────

console.log('\n── scoreRisk ─────────────────────────────────────────────────────');

test('scoreRisk returns 100 for no risk factors', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.scoreRisk([]);
  assert.strictEqual(r.score, 100);
  assert.strictEqual(r.grade, RISK_GRADE.GREEN);
});

test('scoreRisk returns lower score for high-probability impacts', () => {
  const eng = new MonteCarloEngine();
  const r   = eng.scoreRisk([
    { probability: 0.9, impact: 0.9 },
  ]);
  assert(r.score < 30, `Expected low score, got ${r.score}`);
});

// ─── History & Status ─────────────────────────────────────────────────────────

console.log('\n── History & Status ──────────────────────────────────────────────');

test('getHistory returns recent simulations', () => {
  const eng = new MonteCarloEngine();
  eng.runSimulation({ name: 'h1', seed: 1 }, 100);
  eng.runSimulation({ name: 'h2', seed: 2 }, 100);
  const hist = eng.getHistory(5);
  assert(hist.length === 2);
  assert(hist[0].scenario === 'h1');
});

test('clearHistory empties the history', () => {
  const eng = new MonteCarloEngine();
  eng.runSimulation({ seed: 1 }, 100);
  eng.clearHistory();
  assert.strictEqual(eng.getHistory().length, 0);
});

test('status returns totalRuns and phi', () => {
  const eng = new MonteCarloEngine();
  eng.runSimulation({ seed: 1 }, 100);
  const s = eng.status();
  assert.strictEqual(s.totalRuns, 1);
  assert.strictEqual(s.phi, PHI);
});

// ─── Summary ──────────────────────────────────────────────────────────────────

console.log(`\n── Results ──────────────────────────────────────────────────────`);
console.log(`   Passed: ${passed}`);
console.log(`   Failed: ${failed}`);
console.log(`   Total:  ${passed + failed}`);

if (failed > 0) process.exit(1);

File: python/tests/test_monte_carlo.py
"""
Tests for Heady™ Monte Carlo Simulation Engine.
© 2024-2026 HeadySystems Inc. All Rights Reserved.
"""

import pytest

from core.monte_carlo import (
    MonteCarloEngine,
    mulberry32,
    RiskGrade,
    RiskFactor,
    ReadinessSignals,
    ReadinessResult,
    SimulationResult,
    score_to_grade,
)


class TestMulberry32:
    def test_returns_callable(self):
        rand = mulberry32(42)
        assert callable(rand)

    def test_returns_float_in_range(self):
        rand = mulberry32(42)
        for _ in range(1000):
            val = rand()
            assert 0.0 <= val < 1.0

    def test_reproducible_with_same_seed(self):
        rand1 = mulberry32(42)
        rand2 = mulberry32(42)
        for _ in range(100):
            assert rand1() == rand2()

    def test_different_seeds_differ(self):
        rand1 = mulberry32(42)
        rand2 = mulberry32(99)
        # At least some values should differ
        vals1 = [rand1() for _ in range(10)]
        vals2 = [rand2() for _ in range(10)]
        assert vals1 != vals2

    def test_uniform_distribution(self):
        """Values should be approximately uniformly distributed."""
        rand = mulberry32(42)
        vals = [rand() for _ in range(10000)]
        mean = sum(vals) / len(vals)
        assert abs(mean - 0.5) < 0.05  # Should be close to 0.5


class TestRiskGrade:
    def test_grade_enum_values(self):
        assert RiskGrade.GREEN == "GREEN"
        assert RiskGrade.YELLOW == "YELLOW"
        assert RiskGrade.ORANGE == "ORANGE"
        assert RiskGrade.RED == "RED"


class TestScoreToGrade:
    def test_green(self):
        assert score_to_grade(100) == RiskGrade.GREEN
        assert score_to_grade(80) == RiskGrade.GREEN

    def test_yellow(self):
        assert score_to_grade(79) == RiskGrade.YELLOW
        assert score_to_grade(60) == RiskGrade.YELLOW

    def test_orange(self):
        assert score_to_grade(59) == RiskGrade.ORANGE
        assert score_to_grade(40) == RiskGrade.ORANGE

    def test_red(self):
        assert score_to_grade(39) == RiskGrade.RED
        assert score_to_grade(0) == RiskGrade.RED


class TestMonteCarloEngineInit:
    def test_default_seed(self):
        engine = MonteCarloEngine()
        assert engine.default_seed == 42

    def test_custom_seed(self):
        engine = MonteCarloEngine(default_seed=123)
        assert engine.default_seed == 123

    def test_empty_history(self):
        engine = MonteCarloEngine()
        assert engine.get_history() == []


class TestQuickReadiness:
    def test_perfect_signals(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals())
        assert result.score == 100
        assert result.grade == RiskGrade.GREEN

    def test_high_error_rate(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals(error_rate=0.5))
        assert result.score < 100
        assert result.breakdown["error_score"] == 0.0

    def test_failed_deploy(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals(last_deploy_success=False))
        assert result.breakdown["deploy_score"] == 30.0

    def test_high_cpu_pressure(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals(cpu_pressure=0.9))
        assert result.breakdown["cpu_score"] == pytest.approx(10.0)

    def test_zero_service_health(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals(service_health_ratio=0.0))
        assert result.breakdown["health_score"] == 0.0

    def test_multiple_incidents(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals(open_incidents=5))
        assert result.breakdown["incident_score"] == 25.0

    def test_all_bad_signals(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness(ReadinessSignals(
            error_rate=1.0,
            last_deploy_success=False,
            cpu_pressure=1.0,
            memory_pressure=1.0,
            service_health_ratio=0.0,
            open_incidents=10,
        ))
        assert result.grade == RiskGrade.RED

    def test_default_signals(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness()
        assert result.score == 100
        assert result.grade == RiskGrade.GREEN

    def test_returns_readiness_result(self):
        engine = MonteCarloEngine()
        result = engine.quick_readiness()
        assert isinstance(result, ReadinessResult)
        assert "error_score" in result.breakdown


class TestRunFullCycle:
    def test_basic_run(self):
        engine = MonteCarloEngine()
        result = engine.run_full_cycle(name="test", seed=42)
        assert isinstance(result, SimulationResult)
        assert result.scenario == "test"
        assert result.iterations == 10000
        assert result.seed == 42

    def test_reproducible_with_seed(self):
        engine1 = MonteCarloEngine()
        engine2 = MonteCarloEngine()
        r1 = engine1.run_full_cycle(name="test", seed=42)
        r2 = engine2.run_full_cycle(name="test", seed=42)
        assert r1.confidence == r2.confidence
        assert r1.failure_rate == r2.failure_rate

    def test_with_risk_factors(self):
        engine = MonteCarloEngine()
        factors = [
            RiskFactor(name="network", probability=0.3, impact=0.4),
            RiskFactor(name="disk", probability=0.2, impact=0.5, mitigation="RAID"),
        ]
        result = engine.run_full_cycle(name="risky", seed=42, risk_factors=factors)
        assert 0 <= result.confidence <= 100
        assert 0.0 <= result.failure_rate <= 1.0
        assert result.outcomes.success + result.outcomes.partial + result.outcomes.failure == 10000

    def test_mitigation_reduces_failure(self):
        engine = MonteCarloEngine()
        factors_no_mitigation = [
            RiskFactor(name="network", probability=0.5, impact=0.8),
        ]
        factors_with_mitigation = [
            RiskFactor(name="network", probability=0.5, impact=0.8, mitigation="failover"),
        ]
        r_no = engine.run_full_cycle(name="no_mit", seed=42, risk_factors=factors_no_mitigation)
        r_yes = engine.run_full_cycle(name="yes_mit", seed=42, risk_factors=factors_with_mitigation)
        assert r_yes.failure_rate <= r_no.failure_rate

    def test_top_mitigations_populated(self):
        engine = MonteCarloEngine()
        factors = [
            RiskFactor(name="a", probability=0.5, impact=0.5, mitigation="backup"),
            RiskFactor(name="b", probability=0.5, impact=0.5, mitigation="failover"),
        ]
        result = engine.run_full_cycle(name="test", seed=42, risk_factors=factors)
        assert len(result.top_mitigations) <= 5
        assert all(isinstance(m, str) for m in result.top_mitigations)

    def test_confidence_bounds(self):
        engine = MonteCarloEngine()
        result = engine.run_full_cycle(name="test", seed=42)
        assert 0.0 <= result.confidence_bounds.lower <= result.confidence_bounds.upper <= 1.0

    def test_no_risk_factors_all_success(self):
        engine = MonteCarloEngine()
        result = engine.run_full_cycle(name="safe", seed=42, risk_factors=[])
        assert result.confidence == 100
        assert result.failure_rate == 0.0
        assert result.outcomes.success == 10000

    def test_custom_iterations(self):
        engine = MonteCarloEngine()
        result = engine.run_full_cycle(name="small", seed=42, iterations=100)
        assert result.iterations == 100
        assert result.outcomes.success + result.outcomes.partial + result.outcomes.failure == 100


class TestHistory:
    def test_history_populated_after_run(self):
        engine = MonteCarloEngine()
        engine.run_full_cycle(name="test1", seed=42)
        engine.run_full_cycle(name="test2", seed=43)
        history = engine.get_history()
        assert len(history) == 2
        assert history[0].scenario == "test1"
        assert history[1].scenario == "test2"

    def test_history_limit(self):
        engine = MonteCarloEngine()
        for i in range(30):
            engine.run_full_cycle(name=f"test{i}", seed=i, iterations=100)
        history = engine.get_history(limit=5)
        assert len(history) == 5

    def test_status(self):
        engine = MonteCarloEngine()
        status = engine.status()
        assert status["total_runs"] == 0
        assert status["last_run"] is None

        engine.run_full_cycle(name="test", seed=42, iterations=100)
        status = engine.status()
        assert status["total_runs"] == 1
        assert status["last_run"] is not None
